src/backend/distributed/utils/citus_clauses.c

```diff
@@ -146,9 +146,22 @@ PartiallyEvaluateExpression(Node *expression,
 	}
 	else if (nodeTag == T_Query)
 	{
-		return (Node *) query_tree_mutator((Query *) expression,
+		Query *query = (Query *) expression;
+		MasterEvaluationContext subContext = *masterEvaluationContext;
+		if (query->commandType != CMD_SELECT)
+		{
+			/*
+			 * Currently INSERT SELECT evaluates stable functions on master,
+			 * while a plain SELECT does not. For evaluating SELECT evaluationMode is
+			 * EVALUATE_PARAMS, but if recursing into a modifying CTE switch into
+			 * EVALUATE_FUNCTIONS_PARAMS.
+			 */
+			subContext.evaluationMode = EVALUATE_FUNCTIONS_PARAMS;
+		}
```

Comment:
should we explicitly set it in the `else` case to correctly handle regular subqueries and SELECT CTEs?

Reply:
(could also use tests with nested CTEs/subqueries)

Reply:
This should be addressed, though it gets a bit hairy since `INSERT SELECT` evaluates functions in the `SELECT` *(`multi_insert_select` has a test where `evaluate_on_master` behavior varies if you change it from an `INSERT SELECT` to a `SELECT`)*

Reply:
Should we address that first then? Otherwise we end up proliferating the issue.